### PR TITLE
Spell check single files

### PR DESCRIPTION
**Pull request by ThomasHepworth**
### Type of PR

- [ ] BUG
- [x] FEAT
- [ ] MAINT
- [ ] DOC

### Is your Pull Request linked to an existing Issue or Pull Request?
Adds the ability to specify a file or folder path while spellchecking. This builds on top of the infrastructure setup by Zoe in PR https://github.com/moj-analytical-services/splink/pull/1588.

### Give a brief description for the solution you have provided

This pull request:
1. Migrates the dependency checks out of the original spellchecking script so we can use this logic elsewhere, if required.
2. Adds the ability to manipulate the original `.spellchecking.yml` file's source so you can adjust 

On 2, this means you can now point the spellchecker at specific files and folders that will be subsequently checked. Invalid files and folders will spits out errors.

For example: 
```sh
# Checks all docs
source scripts/pyspelling/spellchecker.sh

# Checks only markdown files in the dev guides section of the docs
source scripts/pyspelling/spellchecker.sh docs/dev_guides

# Checks the debug mode file
source scripts/pyspelling/spellchecker.sh docs/dev_guides/debug_modes.md
```

It does this by directly manipulating the source field in our [spellchecking yaml file](https://github.com/moj-analytical-services/splink/blob/master/scripts/pyspelling/.pyspelling.yml#L6).

I'll add some documentation when I'm feeling better.

### PR Checklist

- [ ] Added documentation for changes (TODO)
- [ ] Added feature to example notebooks or tutorial (if appropriate)
- [ ] Added tests (if appropriate)
- [ ] Updated CHANGELOG.md (if appropriate)
- [ ] Made changes based off the latest version of Splink
- [ ] Run the [linter](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/lint.html)

### CODE REVIEW

**Review by RossKen — CHANGES_REQUESTED** [review]
This is a really good functionality to have!

One thing we are missing on the spell checking more generally is documentation - given that it has been place for a while I know I haven't used it (despite spending a lot of time writing docs).

@ThomasHepworth and @zslade I would be keen to know your thoughts. Should we write up a dev guide for it? Or set up a github action to do this (essentially as a docs equivalent to linting)? Or both?

**Comment by zslade** [comment]
Thank you for doing this Tom!

**Comment by zslade** [comment]
> This is a really good functionality to have!
> 
> One thing we are missing on the spell checking more generally is documentation - given that it has been place for a while I know I haven't used it (despite spending a lot of time writing docs).
> 
> @ThomasHepworth and @zslade I would be keen to know your thoughts. Should we write up a dev guide for it? Or set up a github action to do this (essentially as a docs equivalent to linting)? Or both?

Agreed that we need documentation. I think a new section under 'Contributing' called 'Contributing to the Docs' could work where we could put the the info currently in 'Building the Docs' and additional guidance on using the Spellchecker, as two sections.

I have also not used it despite setting up the functionality in the first place (lol). Given that, setting up a GitHub Action seems like a sensible thing to do! Maybe adding another check box to the PR template ('Run the spellchecker if updating the Docs') just as for the linter would also be useful. I find that checkbox for the linter a really helpful reminder to actually run the thing, so hopefully the same will be true of the spellchecker 🔤 ✅

File: scripts/pyspelling/spellchecker.sh [hunk 1]
```diff
@@ -3,22 +3,16 @@
...
-        echo "$line_block Installing ${package_name} $line_block"
-        brew install aspell
-    fi
-else
-    echo "Homebrew is not installed on this system."
-    echo "Please install via https://brew.sh/ and rerun spellchecker.sh"
-    return
-fi
+# Use either the first command line arg or the default path to spellcheck
+path_to_spellcheck="${1:-$default_path_to_spellcheck}"
```
Comment (zslade) [inline scripts/pyspelling/spellchecker.sh:10]: I really like this 👌

File: scripts/utils/ensure_packages_installed.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        return 1
+    fi
+
+    # Iterate over all arguments which are package names
+    for package_name in "$@"; do
+        if brew list --formula | grep -q "^${package_name}\$"; then
+            echo "${package_name} is already installed."
+        else
+            echo "${package_name} is not installed. Installing ${package_name}..."
+            brew install "${package_name}"
```
Comment (zslade) [inline scripts/utils/ensure_packages_installed.sh:18]: Should we install these packages within the venv as well, along with pyspelling?

Comment (ThomasHepworth) [reply scripts/utils/ensure_packages_installed.sh:18]: They're system installs, so I'm not sure we should here.

Comment (ThomasHepworth) [reply scripts/utils/ensure_packages_installed.sh:18]: Is this step still breaking for you if a package isn't installed?

File: scripts/pyspelling/custom_dictionary.txt [hunk 3]
```diff
@@ -45,4 +52,5 @@ tradeoff
 transpilation
 Transpilation
 unlinkables
+winkler
```
Comment (zslade) [inline scripts/pyspelling/custom_dictionary.txt:55]: Should 'winkler' lowercase be included?

File: scripts/utils/ensure_packages_installed.sh (hunk 2 above)
Comment (zslade) [reply scripts/utils/ensure_packages_installed.sh:18]: Not even for system installs that are (for our purposely probably) only going to be used with the spellchecker e.g. `aspell`, which `pyspelling` is built on top of?

Comment (zslade) [reply scripts/utils/ensure_packages_installed.sh:18]: Nope, installs are working fine now 🙌

**Review by zslade — APPROVED** [review]
A couple of comments from me but apart from that, looks good!

File: scripts/pyspelling/custom_dictionary.txt (hunk 3 above)
Comment (ThomasHepworth) [reply scripts/pyspelling/custom_dictionary.txt:55]: I think you're correct... removing this entry.

File: scripts/utils/ensure_packages_installed.sh (hunk 2 above)
Comment (ThomasHepworth) [reply scripts/utils/ensure_packages_installed.sh:18]: If we're using homebrew to install these, I'm not sure we can.

If we could, we'd need to make an additional os environment as these can't be installed in the python venv - I'm not sure it's worth the faff of setting up (or if it's possible outside of docker).

**Pull request merged**